### DISCUSSION TITLE
@CustomType

### DISCUSSION THREAD

**Discussion by sebersole** [hibernate/hibernate-orm#4266] [Design Proposals]
TLDR; Introduce `@CustomType` as a type-safe replacement for `@Type`.

https://hibernate.atlassian.net/browse/HHH-14856

We've had a few discussions now about wanting to move away from Type determination being based on Strings and to define a more type-safe way to accomplish that.

I propose the following annotation:

```
public @interface CustomType {
	/**
	 * The attribute "name" to which the type applies.  The
	 * "name" can also specify that it applies to the key
	 * of the Map attribute, etc.
	 *
	 * We recognize the standard Jakarta Persistence values, when applied Map attributes:<ul>
	 *     <li>{@code "key"}</li>
	 *     <li>{@code "value"}</li>
	 * </ul>
	 */
	String attributeName() default "";

	/**
	 * The custom type implementation
	 */
	Class<? extends UserType<?>> typeImplementation();

	/**
	 * Parameters to be injected into the custom type implementation
	 * after it is instantiated. To receive the parameters, the custom
	 * type implementation must implement
	 * {@link org.hibernate.usertype.ParameterizedType}
	 */
	Parameter[] parameters() default {};
}
```

With regard to `#attributeName`, it was suggested by @beikov and I think its a great idea.  Rather than needing additional annotations such as `@MapKeyCustomType`, etc we can just have a single annotation and specify what the type applies to.  This follows the pattern used by JPA so that's good(ish)

I also propose that we drop `@Type`, `@MapKeyType`, `@TypeDef`, `@TypeDefs` as part of this work.  We have better ways for users to accomplish these, so there is good migration story.  Keeping `UserType` and providing `@CustomType` covers the remaining cases where people want/have to continue to use UserType impls.  And it gets us out of the String-based resolution at all.  Well, from annotations at least.

**Comment by sebersole** [comment]
`org.hibernate.annotations.AnyMetaDef` also uses "type name" : `#metaType` and `#idType`

**Comment by yrodiere** [comment]
What would be the solution for automatically applying a custom type to a given class? We used to have `@TypeDef#defaultForType`, but if you remove `@TypeDef`...

Thinking about it, shouldn't the solution be similar to `@Converter(autoApply = true)`? I.e. ask users to just apply an annotation on their custom type, and detect that during classpath scanning?

**Comment by sebersole** [comment]
Depending on exactly what the custom UserType did...

1. Using `AttributeConverter` would be the first suggestion.  Its good for what it does, though not "complete" - e.g. it misses anything related to what we call `MutabilityPlan`.  And yes, it can be auto-discovered.
2. A custom `@JavaTypeRegistration` which can also be supplied by `TypeContributor`, `Dialect`, etc.  `TypeContributor` is also auto-detected.

**Comment by gavinking** [comment]
@sebersole I don't think I like this `attributeName` thing, but I think maybe you already moved away from that, is that right?

**Comment by sebersole** [comment]
No I like it.  It helps solve the fugly annotations-dont-have-inheritance-so-we-need-to-wrap-the-annotation-we-are-interested-in-within-other-annotations.  In other words, I find this nicer 

```
@CustomType( typeImplementation = MyUserType.class )
@CustomType( typeImplementation = AnotherUserType.class, attributeName = "key" )
Map someMap;
```

than

```
@CustomType( typeImplementation = MyUserType.class )
@MapKeyCustomType( CustomType( typeImplementation = AnotherUserType.class ) )
Map someMap;
```

at least from the perspective of implementing support for it.  In my opinion its arguable which is better from a user perspective and some will prefer one while others prefer the other.  But the one argument for it as-is is consistency with JPA e.g. with `@Convert`, `@AttributeOverride`, etc

And keep in mind, its not just `@MapKeyCustomType` that needs this.. we also have 

* `@AnyMetaDef#metaType`
* `@AnyMetaDef#idType`
* `@CollectionId#type`
* `@ParamDef#type`

**Reply by gavinking** [reply]
Well OK but then let's use an enum instead of strings.

**Reply by sebersole** [reply]
That's a good compromise, though I do wonder if we want to allow this, e.g., for "embeddable overrides"...

**Reply by beikov** [reply]
That's the idea behind the string path for the JPA `@Convert` annotation at least.

**Reply by gavinking** [reply]
String _paths_ are especially evil. They break the Rename refactoring, of which I'm an habitual/addicted user.

**Reply by gavinking** [reply]
(Of course, we already have them all over the place in queries and also in `mappedBy`. But I *hate* `mappedBy` for this reason.)

**Reply by sebersole** [reply]
So you'd vote for dedicated annotations for each usage instead?  Aka:

* `@CustomType`
* `@MapKeyCustomType`
* `@AnyKeyCustomType`
* `@AnyDiscriminatorCustomType`
* `@CollectionIdCustomType`
* `@DiscriminatorCustomType`
* ...
?

**Reply by sebersole** [reply]
And even worse actually...

```
@interface CustomType { ... }
@interface MapKeyCustomType { CustomType value() }
...
```

**Reply by gavinking** [reply]
I mean, sure, that's a *lot*. And I guess there's no way to hang those off of other annotations? e.g. `@Discriminator(customType=@CustomType(MyType.class))`, `@MapKey(customType=@CustomType(MyType.class))`.

**Reply by sebersole** [reply]
Its programming.. we can do whatever we want.  So sure, that is possible.  But if we are going down that rabbit hole, then we should really combine all the other "metadata" about the specific things into the "hung-off-of" annotations:

For example, we already have `@CollectionId` so that one is easy enough - so let's start there

```
@interface CollectionId {
    Column[] columns();
    String generator();

    // Type type();
    CustomType type();
}
```


But then do we do the same for `@JavaType`, `@JdbcType`, ... related to collection-id (which I forgot to add support for anyway)?  E.g.

```
@interface CollectionId {
    Column[] columns();
    String generator();

    // Type type();
    CustomType type();

    JavaType javaType();
    JdbcType jdbcType();
    JdbcTypeCode jdbcTypeCode();
    ...
}
```
?

Logically inconsistent to not do that imo.

As we keep saying, this is a major release and is the appropriate time to fix things where we made poor decisions in the past.  In my opinion annotations were never defined with an overall idea or consistent pattern in mind and we should address those if they apply here.

**Reply by sebersole** [reply]
And to be clear - I HATE that last example

**Reply by gavinking** [reply]
> we should really combine all the other "metadata" about the specific things into the "hung-off-of" annotations:

Right, that's what I'm suggesting.

> But then do we do the same for `@JavaType`, `@JdbcType`

Well I really do _much_ prefer `@CollectionId(javaType=@JavaType(MyDescriptor.class))` or even just `@CollectionId(javaType=MyDescriptor.class)` to `@JavaType(attributeName="collectionId", descriptor=MyDescriptor.class)`.

**Reply by sebersole** [reply]
for each basic-valued mapping you can specify

1) custom-type
2) java-type descriptor
3) jdbc-type descriptor
4) jdbc code (to resolve the descriptor)

in terms of determining the mapping (ignoring AttributeConverter and mutability for now).  So following your example, you'd have:

```
@interface CollectionId {
    Column[] columns();
    String generator();

    CustomType type();

    JavaType javaType();
    JdbcType jdbcType();
    JdbcTypeCode jdbcTypeCode();
    ...
}
```

now obviously some of these are mutually exclusive, some optional, ...  I'm sorry, but I think that is just terrible design.

Your's looks simpler because you ignore all of the possible options

**Reply by gavinking** [reply]
> now obviously some of these are mutually exclusive, some optional

Well of course in practice you'll only ever specify one of them. And sure, it sucks that we can't impose that as a compile-time constraint.

But we can't impose the constraint with any of the other competing options either, so I'm unclear on how it counts against this option. 

At least with this approach things that belong together are grouped naturally together. 

Imagine how much it would suck if you had to specify all the field type mappings for an entity at the top of the class and point to the fields in `attributeName`. That would be awful. Sadly we don't have a field on which to place collection id type mappings, but it seems to me that the `@CollectionId` annotation is the next best thing to that.

**Reply by beikov** [reply]
We could allow `@Target(TYPE_USE)` so that one can do `Map<@Type( MyKeyType.class ) MyKey, @Type( MyValueType.class ) MyValue> myMap;`

**Reply by gavinking** [reply]
Oh wow, that's a great thought!

Are `TYPE_USE` annotations available via reflection?

**Reply by sebersole** [reply]
So just for fun, I started breaking down all of the basic-value mapping annotations into this pattern.  For the moment, let's assume we define specific annotations for each "kind" (attribute, map-key, collection-id, etc) and let's assume that we don't use nested annotations.  That still means we need:

```
@interface BasicMapping {
    Class<? extends UserType> customType() default void.class;
    Class<? extends JavaTypeDescriptor> javaType() default void.class;
    Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan> mutability() default void.class;
}

@interface MapKeyMapping {
    
    Class<? extends UserType> customType() default void.class;
    Class<? extends JavaTypeDescriptor> javaType() default void.class;
    Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan> mutability() default void.class;    ​
}

// we'll assume that `@BasicMapping` applied to a Map describes the value
// and we'll assume that `@BasicMapping` applied to a Collection describes the element

@interface CollectionIdMapping {

    /** Not sure why we define column*s* here.  We do not support composite id-bag ids */
    Column[] columns();
    String generator();


    Class<? extends UserType> customType() default void.class;
    Class<? extends JavaTypeDescriptor> javaType() default void.class;
    Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan> mutability() default void.class;    ​
}


/** Completely re-work annotations for ANY mappings... they are awful as-is */
@interface AnyMapping {
    Column metaColumn();
    FetchType fetch() default FetchType.EAGER;
    ...
    
    Discriminator discriminator();

    @interface Discriminator {
        Column column();
        Class<? extends UserType> customType() default void.class;
        Class<? extends JavaTypeDescriptor> javaType() default void.class;
        Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
        int jdbcTypeCode() default Integer.MIN_VALUE;
        Class<? extends MutabilityPlan> mutability() default void.class;
    }

    // key / id can be composite, so not sure the best way to map that using this approach
}

@interface ParamDef {
	String name();
    Class<? extends UserType> customType() default void.class;
    Class<? extends JavaTypeDescriptor> javaType() default void.class;
    Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan> mutability() default void.class;    ​
}

```

and that has not even accounted for embeddable overrides.  So we'd also need:

```
@interface BasicMappingOverride {
    String name();

    // though here, nested might be ok - but consistency...
    Class<? extends UserType> customType() default void.class;
    Class<? extends JavaTypeDescriptor> javaType() default void.class;
    Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan> mutability() default void.class;    ​
}

@interface MapKeyMappingOverride {
    String name();

    Class<? extends UserType> customType() default void.class;
    Class<? extends JavaTypeDescriptor> javaType() default void.class;
    Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan> mutability() default void.class;    ​
}

...

```

I mean that's logically consistent right?

**Reply by sebersole** [reply]
> But we can't impose the constraint with any of the other competing options either, so I'm unclear on how it counts against this option.

Because we have one annotation in the other option, and n+1 here.  N+1 = bad :laughing:

**Reply by sebersole** [reply]
> Imagine how much it would suck if you had to specify all the field type mappings for an entity at the top of the class and point to the fields in `attributeName`. That would be awful. 

No one suggested that.  I think you might be mixing discussions.

**Reply by sebersole** [reply]
> Sadly we don't have a field on which to place collection id type mappings, but it seems to me that the `@CollectionId` annotation is the next best thing to that.

Not understanding this one.  You mean something different than:

```
@CustomType( attributeName = COLLECTION_ID, typeImplementation = ... )
Collection myIdBag;
```

?

**Reply by gavinking** [reply]
> Because we have one annotation in the other option, and n+1 here. N+1 = bad 

I don't think that's quite right. I think that for each of the mappings that you could specify a custom type, you could also specify other things. For example, `@CollectionId` already exists.

Now, cutting against that is the fact that in certain cases, e.g. `@MapKey`, the annotation isn't under our control, and so we can't add attributes to it, and that's a pain, because it would require something "new", perhaps.

> No one suggested that.

It's an analogy. Analogies aren't strawmen, at least not in general.

**Reply by gavinking** [reply]
> Not understanding this one.

Go back to my analogy. It would suck if we mapped entities like this:

```java
@Entity
@Column(attributeName="id", name="person_id")
@Column(attributeName="name", name="full_name")
@CustomType(attributeName="name", type=StringType.class)
@CustomType(attributeName="address", type=AddressType.class)
class Person {
}
```

Luckily for us, we have convenient "slots" for those annotations (the Java fields).

So what I'm suggesting is that for those things for which we don't have "slots", we create a slot using an annotation. So things which "annotate the collection id field" go in `@CollectionId`.

That's intellectually consistent.

**Reply by sebersole** [reply]
> > No one suggested that.
> 
> It's an analogy. Analogies aren't strawmen, at least not in general.

:shrug: You said it, so I responded.  Intentions don't jump off the page or screen ;)


> I don't think that's quite right. I think that for each of the mappings that you could specify a custom type, you could also specify other things. For example, `@CollectionId` already exists.

Ok, then I amend my earlier novel to now include:

```
@interface BasicMapping {
    Class<? extends UserType> customType() default void.class;
    Class<? extends JavaTypeDescriptor> javaType() default void.class;
    Class<? extends JdbcTypeDescriptor> jdbcType() default void.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan> mutability() default void.class;

    /** And drop @Immutable */
    boolean immutable() default false;

    /** And drop @Check */
    String checkConstraints() default "";

    String comment();

    ...
}
```

I know you aren't suggesting that, but that's the logical analogy to what you are saying ;)

**Reply by gavinking** [reply]
P.S. on a minor point @sebersole you wrote:

    @CustomType( typeImplementation = MyUserType.class )

I think `typeImplementation` really should be `value`, otherwise you're disrupting the overwhelmingly most common case which is the use of these annotations on Java fields. For this common case I want to write `@JavaType( BitSetJavaType.class )` not `@JavaType( typeDescriptor = BitSetJavaType.class )`.

**Reply by gavinking** [reply]
@sebersole I don't understand the purpose of `@BasicMapping`. 

If it's meant for regular fields, then it's unnecessary. Regular fields are already a "slot" where I can put annotations. I'm not sure if you're understanding my whole point here.

**Reply by gavinking** [reply]
My point is that: 

> For things that **aren't** fields, we don't have a place to put annotations.

And that's the root source of the discomfort.

So create such a place.

**Reply by beikov** [reply]
>Are TYPE_USE annotations available via reflection?

Yes, so that will help with the sane cases. If users want to use raw types, they still can use the attribute/path approach

**Reply by sebersole** [reply]
> P.S. on a minor point @sebersole you wrote:
> 
> ```
> @CustomType( typeImplementation = MyUserType.class )
> ```
> 
> I think `typeImplementation` really should be `value`, otherwise you're disrupting the overwhelmingly most common case which is the use of these annotations on Java fields. For this common case I want to write `@JavaType( BitSetJavaType.class )` not `@JavaType( typeDescriptor = BitSetJavaType.class )`.

Yes.  But that was just copy/paste from the form defining attributeName as well.  Yes, if there is just one value I'd use value.  Ofc then your risk the problem of needing to add an attribute later and then value is (most likely) no longer the right name.  

But I agree in that I would do the same with just one attribute

**Reply by sebersole** [reply]
> My point is that:
> 
> > For things that **aren't** fields, we don't have a place to put annotations.
> 
> And that's the root source of the discomfort.
> 
> So create such a place.

Its actually possible to understand what you are saying and yet still disagree with you ;)

I understand exactly what you are saying.  

Personally I hate these TYPE_USE annotations generally speaking.  I think they make code way harder to read.  But its an elegant solution for sure

**Reply by sebersole** [reply]
> > Are TYPE_USE annotations available via reflection?
> 
> Yes, so that will help with the sane cases. If users want to use raw types, they still can use the attribute/path approach

Minor point, but there will never be such a place for a collection-id.

**Reply by sebersole** [reply]
Nor ANY mappings

**Comment by sebersole** [comment]
As I played with prototyping this (and porting tests, especially) I wonder if it is worthwhile to still provide a `@TypeDef`-like annotation, though limited in its scope to just the "auto-apply" aspect of TypeDef (no "name").  In other words, I think this might be worth-while:

```
@Repeatable( CustomTypeRegistrations.class )
public @interface CustomTypeRegistration {
	Class<? extends UserType<?>> typeImplementation();
	Class<?>[] autoApplyTo() default {};
	Parameter[] parameters() default {};
}
```

As well as adjusting `TypeContributions`:

```
interface TypeContributions {
    ...
    void contributeType(Class<? extends UserType> impl, Class<?>... autoApplyTo);
}
```

Coming back to the enum point.. as I mentioned, I think it depends on what we want to support.  2 things in particular I'd like to support that we need to consider here:

1) overrides for embeddable values (think `@AttributeOverride`, etc)
2) applying this for "entity metadata" - discriminator mostly

And of course it depends how we want to support those.  

For discriminator, e.g, do we add a `@DiscriminatorCustomType` annotation?  That kind of defeats the purpose of having a singular annotation for applying custom-types.  On the other hand, the alternative is to allow using `@CustomType` to the entity/mapped-super which makes it confusing - then certain "paths" are valid based on where it is used and `#attributeName` is required.  

For overrides, the difference comes back to needing to supply open-ended values (i.e. actual attribute names)

Don't know the answers yet; just some thoughts that ran through my head as I started implementing this

**Reply by gavinking** [reply]
My only concern with stuff like `@CustomTypeRegistration` though it's *by no means* specific to this annotation, and applies to others that we already have, is: where do I put this annotation that properly reflects to scope to which it applies?

**Reply by sebersole** [reply]
> My only concern with stuff like `@CustomTypeRegistration` though it's _by no means_ specific to this annotation, and applies to others that we already have, is: where do I put this annotation that properly reflects to scope to which it applies?

I'd limit those to just package and class and say they are global.  It's really just an easy form of using TypeContributor

**Reply by beikov** [reply]
If it is just "sugar", I'd opt to not having the annotation at all and require a TypeContributor instead.

**Reply by sebersole** [reply]
Yes, it is "sugar". But sugar that is very useful for tests. Unless we add something like @DomainModel( ... typeContributors=... )

Which is fine as well.

**Reply by sebersole** [reply]
My concern there though is that if our tests need it, it is likely users will also. But I am fine aggressively and proactively defining what we want here and maybe needing to go back and add something like this if enough users complain

**Comment by sebersole** [comment]
So I changed my mind a bit.  I (and I think all of us) would preferred annotations to be able to implement a common interface, but it ofc does not.  So the next best thing is to use the nested annotation approach.  We can look at the type-use approach later as a possible improvement.  So I'll do:

```
@interface CustomType {
    Class<? extends UserType> value();
    Parameter[] parameters() default [];
}

@interface MapKeyMapping {
    CustomType customType() default ...;
    JavaType javaTypeDescriptor() default ...;
    JdbcType jdbcTypeDescriptor() default ...;
    JavaTypeCode jdbcTypeCode() default ...;
    Mutability mutabilityPlan() default ...;
    // Could just rely on JPA's {@link Convert#attributeName},
    // but I like it being available here as well
    Class<? extends AttributeConverter> converter();
}

/** Could just add to `@CollectionId` instead
@interface CollectionIdMapping {
    CustomType customType() default ...;
    JavaType javaTypeDescriptor() default ...;
    JdbcType jdbcTypeDescriptor() default ...;
    JavaTypeCode jdbcTypeCode() default ...;
    Mutability mutabilityPlan() default ...;
    // Could just rely on JPA's {@link Convert#attributeName},
    // but I like it being available here as well
    Class<? extends AttributeConverter> converter();
}

...
```

Though, even this might be better to allow for adding addition mapping details

```
@interface CustomType {
    Class<? extends UserType> value();
    Parameter[] parameters() default [];
}

@interface BasicMapping {
    CustomType customType() default ...;
    JavaType javaTypeDescriptor() default ...;
    JdbcType jdbcTypeDescriptor() default ...;
    JavaTypeCode jdbcTypeCode() default ...;
    Mutability mutabilityPlan() default ...;
    // Could just rely on JPA's {@link Convert#attributeName},
    // but I like it being available here as well
    Class<? extends AttributeConverter> converter();
}

@interface BasicMapKey {
    BasicMapping mapping() default @BasicMapping;
    ...
}

/** 
@interface BasicMapKey {
    BasicMapping mapping() default @BasicMapping;
    ...
}

@interface CollectionId {
    ...
    BasicMapping mapping() default @BasicMapping;
    ...
}

// etc
```

**Reply by sebersole** [reply]
I'll send a PR when I am done, so we can fix any of this we need to later, but I am wondering if anyone has improvements to these javadocs so far:

```
/**
 * Describes the mapping of a basic value.  Supports 2 main approaches for adjusting the
 * default mapping Hibernate will use:<ul>
 *     <li>
 *         (the preferred approach) using composition - <ul>
 *             <li>{@link #javaType()}</li>
 *             <li>{@link #jdbcType()}</li>
 *             <li>{@link #jdbcTypeCode()}</li>
 *             <li>{@link #mutability()}</li>
 *             <li>{@link #immutable()}</li>
 *             <li>{@link #converter()}</li>
 *         </ul>
 *     </li>
 *     <li>
 *         contracted via the {@link UserType} interface and {@link #customType()}
 *     </li>
 * </ul>
 *
 * @see jakarta.persistence.Basic
 */
public @interface BasicMapping {
	/**
	 * Defines the {@link org.hibernate.type.descriptor.java.BasicJavaDescriptor} to use.
	 *
	 * To apply this descriptor as the default for all references of this Java type in
	 * the domain model, use {@link JavaTypeRegistration} or
	 * {@link org.hibernate.boot.model.TypeContributor} instead
	 *
	 * @apiNote Mutually exclusive with specifying {@link #customType()}
	 * @implSpec The supplied Class is resolved as a {@link org.hibernate.resource.beans.spi.ManagedBean}
	 */
	JavaType javaType() default @JavaType( NoJavaTypeDescriptor.class );

	/**
	 * Defines the {@link org.hibernate.type.descriptor.jdbc.JdbcTypeDescriptor} to use
	 *
	 * To apply this descriptor as the default for all resolutions of a particular
	 * type-code (see {@link #jdbcTypeCode()}) within the SessionFactory
	 *
	 * @apiNote Mutually exclusive with specifying {@link #jdbcTypeCode()} or {@link #customType()}
	 * @implSpec The supplied Class is resolved as a {@link org.hibernate.resource.beans.spi.ManagedBean}
	 */
	JdbcType jdbcType() default @JdbcType( NoJdbcTypeDescriptor.class );

	/**
	 * Defines the code the is then used to resolve the
	 * {@link org.hibernate.type.descriptor.jdbc.JdbcTypeDescriptor} to use from the
	 * {@link org.hibernate.type.descriptor.jdbc.spi.JdbcTypeDescriptorRegistry} which
	 * is a configurable mapping of a type-code to a descriptor to use
	 *
	 * @apiNote Mutually exclusive with specifying {@link #jdbcType()} or {@link #customType()}
	 */
	JdbcTypeCode jdbcTypeCode() default @JdbcTypeCode( Integer.MIN_VALUE );

	/**
	 * Defines the {@link org.hibernate.type.descriptor.java.MutabilityPlan} to
	 * apply to the mapping.
	 *
	 * @see Immutable
	 *
	 * @apiNote Mutually exclusive with specifying {@link #immutable()} or {@link #customType()}
	 * @implSpec The supplied Class is resolved as a {@link org.hibernate.resource.beans.spi.ManagedBean}
	 */
	Mutability mutability() default @Mutability( NoMutabilityPlan.class );

	/**
	 * By default, basic value mapping are considered mutability (most anyway, depending
	 * on usage site).  This indicates to treat the value as immutable.  Simplified
	 * alternative to {@link #mutability()} for the common case.
	 *
	 * @apiNote Mutually exclusive with specifying {@link #mutability()} or {@link #customType()}
	 */
	boolean immutable() default false;

	/**
	 * Defines a conversion to be applied to the values
	 *
	 * @apiNote Mutually exclusive with specifying {@link #customType()}
	 * @implSpec The supplied Class is resolved as a {@link org.hibernate.resource.beans.spi.ManagedBean}
	 */
	Class<? extends AttributeConverter> converter() default NoAttributeConverter.class;

	/**
	 * Defines a {@link UserType} implementation to use for mapping the values
	 *
	 * @apiNote Mutually exclusive with specifying {@link #javaType()}, {@link #jdbcType()},
	 * 		{@link #jdbcTypeCode()}, {@link #mutability()}, {@link #immutable()} and {@link #converter()}
	 * @implSpec The supplied Class is resolved as a {@link org.hibernate.resource.beans.spi.ManagedBean}
	 */
	CustomType customType() default @CustomType( NoUserType.class );
}

```

**Reply by yrodiere** [reply]
LGTM.

It's interesting, because I think the use case seems somewhat similar to what we have in Hibernate Search, so I'll go ahead and suggest an API similar to what we did in Hibernate Search 6. If that's not relevant, that's fine; I just want to bring it to your attention in case you like it.

I'm assuming that all those options you are providing (`javaTypeDescriptor`, `jdbcTypeDescriptor`, `mutabilityPlan`, etc.) only have a limited amount of valid combinations. If that's wrong, please ignore the rest of this message, as the solution I'm about to propose probably won't be a good fit.

In Hibernate Search, we have a concept of "bridge" and "binder". Both are user-implemented callbacks:

* The "binder" gets called during the mapping phase. It gets access to metadata about the property being "bound", and is able to set metadata about the resulting "column". In your case that metadata would be the `javaTypeDescriptor` (... I guess?), `jdbcTypeDescriptor`, `mutabilityPlan`, and more if you need to.
* The bridge gets called at runtime to perform all operations related to read/write. In your case, I believe that would be the `CustomType`?

What I'm suggesting is, instead of exposing lots of options in your annotations, you could just provide some "binder" API:

```java
@BasicBinding(binder = MyBinder.class)
MyType myProperty;
```

```java
public class MyBinder implements CustomTypeBinder {
    public void bind(CustomTypeBindingContext context) {
        context.mutabilityPlan( MutabilityPlan.SOMETHING );
        context.jdbcTypeDescriptor( SomeJdbcTypeDescriptor.INSTANCE );
        context.customType( new MyCustomType() );
    }

    private static class MyCustomType implements CustomType {
        // ...
    }
}
```

This gives a lot of flexibility. The binder could inspect the property and assign a different custom type depending on the property type, for example:

```java

public class MyBinder implements CustomTypeBinder {
    public void bind(CustomTypeBindingContext context) {
        if ( context.propertyRawType() == String.class ) {
            context.mutabilityPlan( MutabilityPlan.SOMETHING );
            context.jdbcTypeDescriptor( SomeJdbcTypeDescriptor.INSTANCE );
            context.customType( new MyCustomType() );
        }
        else {
            context.mutabilityPlan( MutabilityPlan.SOMETHING_ELSE );
            context.jdbcTypeDescriptor( SomeOtherJdbcTypeDescriptor.INSTANCE );
            context.customType( new MyOtherCustomType() );
        }
    }

    private static class MyCustomType implements CustomType {
        // ...
    }
    private static class MyOtherCustomType implements CustomType {
        // ...
    }
}
```

Another advantage, though it's in no way necessary if you don't like it, is that you're able to move operations out of `CustomType` and into multiple interfaces, each dedicated to a certain aspect of reading/writing. You could have a dedicated interface for `toXMLString`/`fromXMLString`, a dedicated interface for mutability (with `isDirty` + `deepCopy` + any method that only makes sense for mutable types), etc. It helps create an API with more cohesive interfaces.

Here's the documentation of binders/bridges in Hibernate Search, though I don't think it will be very useful to you, since the `context` operations are not likely to be relevant to Hibernate ORM (they're mostly about defining index fields):

* https://docs.jboss.org/hibernate/search/6.1/reference/en-US/html_single/#mapper-orm-bridge-valuebridge-valuebinder
* https://docs.jboss.org/hibernate/search/6.1/reference/en-US/html_single/#mapper-orm-bridge-propertybridge

**Reply by sebersole** [reply]
It's a bit difference.  All of them affect the resolution of mapping details during boot.  At runtime they are all used when actually reading (not yet writing always) values of the particular mapping

**Comment by sebersole** [comment]
Going to make a change as I got into implementing this...

First, I am going to drop:

* `@JavaType`
* `@JdbcType`
* `@JdbcTypeCode`
* `@MutabilityPlan`
* `@MapKeyJavaType`
* `@MapKeyJdbcType`
* `@MapKeyJdbcTypeCode`

Then define `@BasicMapping` as follows (with mostly the same Javadoc as before):

```
@interface BasicMapping {
    Class<? extends BasicJavaDescriptor<?>> javaType() default NoJavaTypeDescriptor.class;
    Class<? extends JdbcTypeDescriptor<?>> jdbcType() default NoJdbcTypeDescriptor.class;
    int jdbcTypeCode() default Integer.MIN_VALUE;
    Class<? extends MutabilityPlan<?>> mutability() default NoMutabilityPlan.class;
    Class<? extends AttributeConverter> converter() default NoAttributeConverter.class;
    CustomType customType() default @CustomType( NoUserType.class );
}
```

And then:

```
@interface BasicMapKey {
    BasicMapping mapping() default @BasicMapping();
    ...
}

@interface CollectionId {
    BasicMapping mapping() default @BasicMapping();
    ...
}

...
```
